### DISCUSSION TITLE
How can I add a column that's a shift from another column?

### DISCUSSION THREAD

**Discussion by jabellcu** [saulpw/visidata#2608] [Q&A] [answered]
I have a column with dates: "Date". I would like to create a new column, "PrevDate" with each row's previous Date value. I have tried to add an expression column with `=` and `Date.shift()`, but that doesn't work. Ultimately, I would like to add a column with the duration between two consecutive rows. Is there a way to acomplish this? Many thanks.

**Comment by midichef** [comment, marked as answer]
Hi, to use the values of rows that are before or after the current row, use the `addcol-window` command, which is bound to `w`.

For your example:  move the cursor to be in the column "Date", then `w` then `1 0`, and then `=` and `Date_window[1] - Date_window[0]`

**Reply by jabellcu** [reply]
This works, thank you very much. I didn't know about `addcol-window`. Still, is there a way to add a column applying pandas functions to other columns as `pd.Series` ie. `=` and then OtherColumn.shift() ?

**Reply by midichef** [reply]
I don't use pandas, so I'm unfamiliar with the details of how `Series` works.

Yes, it can be done, but it is a bit inconvenient. I'll give some example commands for a column named `test`.
Navigate to the test column, then execute: 
`Space` `exec-python` `vals = [vd.sheet.cursorCol.getTypedValue(r) for r in sheet.rows]`
That makes a Python list with the values in the column, accessible as `vals`.
Then create the pandas series:
`Space` `exec-python` `s = pandas.Series(vals).shift()`
Then, to turn it into a column:
`za` `test_shifted`
`gs`
`Space` `setcol-iter` `s`

(update:  I edited to simplify the command sequence, using `setcol-iter` instead of `addcol-incr` and `=`)
(update:  I changed the command sequence to no longer use `z+` `list`)

**Reply by jabellcu** [reply]
Great answer, thank you. I was under the impression visidata's tables were pandas dataframes (and columns would then be series objects).

**Reply by midichef** [reply]
The answer I gave you was flawed. I've edited it to fix it.

The flaw is that it would give bad results if the starting column had any cells that held null, or contained errors. By errors, I mean  the kind of Exception-related errors that would be shown by`error-cell`.

The reason is, `z+` `list` ignores any cells that were null or had errors. So `vd.memory['test_list']` would have fewer elements than needed to fill in the rows of `test_shifted`. The rows at the end would be filled in by `setcol-iter` using repeated values from the start of `vd.memory['test_list']`.

The edited version, using `vals = [vd.sheet.cursorCol.getTypedValue(r) for r in sheet.rows]`, fixes this problem.
